### ISSUE TITLE
Give a better meaning to `--force`

### ISSUE THREAD

**Issue by jeremiedimino** [ocaml/dune#4390]
Currently `--force` only re-execute actions that are part of a `rule` stanza with an `alias` field but no targets. This feels arbitrary.

I propose the following interpretation instead: `--force` re-execute all actions written in `dune` files. This seems less ad-hoc and is probably what users want anyway.

**Comment by rgrinberg** [comment]
The main use case for `--force` is to re-run tests. This is mostly needed because aren't used to dune's feature of not re-running tests when it's not needed. So while the proposed interpretation of `--force` makes sense, it does not seem like it's fulfilling a need for users.

**Comment by jeremiedimino** [comment]
Indeed. Once we have goals, we could change the meaning of `--force` to: re-execute all actions attached to a goal. It would keep the behaviour users care about and would have a simpler definition than the current one.

**Comment by jonsterling** [comment]
Btw `dune runtest --force` doesn't actually rerun my tests! Did I do something wrong? My repo is at http://www.github.com/redprl/cooltt. It couldn't be more frustrating, we are about to switch back to a dumb bash script.

**Comment by ejgallego** [comment]
@jonsterling this is exactly the case Jeremie mentions in this issue, force will only re-run aliases without a target, so what happens when you use force indeed is that the following is re-run [as you can check in _build/log]
```
$ (cd _build/default/test && /bin/bash -e -u -o pipefail -c 'rm ./*.cooltt')
```
however, the output file is still considered up-to-date.

You could either indeed delete that file, and have the tests re-run, or properly setup the deps for your tests so they are run when they are supposed to. IMHO this pays off, as then you don't need to keep that bit of mental state to determine when you need to use `--force`.

**Comment by jonsterling** [comment]
Can you explain what I should do to setup the deps for the tests so that they are "run when they are supposed to"? Btw I want to clarify that I would consider it incorrect behavior for tests to _ever_ not be run because they've "already been run". I think that most people who run tests while developing would agree --- we want to run them again not because they will give a new result, but because we lost the output and want to see it again.

**Comment by jonsterling** [comment]
Btw thanks for your help!

**Comment by ejgallego** [comment]
> I would consider it incorrect behavior for tests to _ever_ not be run because they've "already been run"

Umm, tests are not special in dune, they are just implemented as a (purely functional) build rule. Keep in mind that incremental running of tests is important for some projects as they may have thousand of them.

So if the test rules are correctly setup, tests should be re-run only when some of their deps have changed, otherwise, why you want to re-run a deterministic function that you have already run with the same inputs?

> but because we lost the output and want to see it again.

Oh, in this case IMHO you don't want to re-run the tests, but a simple rule that would display them, something with `(action (cat test-output.out)` , in this case dune will check if test-output.out is up-to-date, and do the right thing; it could be even pulled from the cache.

> Can you explain what I should do to setup the deps for the tests so that they are "run when they are supposed to"?

The setup seems actually close to correct to me, you depend on the test executable and all the test files that are gonna be the input, tho you could tweak stuff a bit, such as:

```
(rule
 (target test.output)
 (deps
  (glob_files ./*.cooltt)
  (glob_files ./cooltt-lib))
 (action (with-stdout-to %{target} (run ./test.exe))))
```
tho I am not sure I follow the what is the purpose of `(bash "rm ./*.cooltt")` , that seems a bit strange?
```

**Comment by jonsterling** [comment]
Well, there is another reason why I really do want to re-run the tests. I may be trying to check how long they take to run. Anyway, I understand why you feel that the behavior of dune is correct (it makes sense in the framework of dune), but I have to insist that it doesn't really match the way that tests are used by engineers. This suggests to me that testing may benefit from a different treatment, given that to obtain the behavior that is expected by me, it seems like one must write some kind of surprising rules in the dune code. I apologize that I am not deeply familiar with dune's architecture, so what seems like an easy-to-fix thing for you is for me something very hard.

**Comment by nojb** [comment]
> Well, there is another reason why I really do want to re-run the tests. I may be trying to check how long they take to run. In general, I can continue coming up with example use-cases that prove (to me) that test output should never be cached in practice for most use-cases.

Can't you use the `(universe)` dependency to avoid the caching?

**Comment by esope** [comment]
Does anybody know whether some progress was made on this issue?

**Comment by rgrinberg** [comment]
No progress was made.
